### PR TITLE
Chapter 7: Sign Up

### PR DESCRIPTION
**Pull request by baaohuii**

### CODE REVIEW

**Comment by baaohuii** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -60,3 +60,70 @@ h2 {
...
+    margin-top: 20px;
+    &:first-child {
+      border: 0;
+      padding-top: 0;
+    }    
+    span {
+      display: block;
+      margin-bottom: 3px;
+      line-height: 1;
+    }
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: image

File: app/assets/stylesheets/users.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/users.scss]: remove file that never used

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  def show
+  	@user = User.find_by id: params[:id]
+  end
+
+  def new
+  	@user = User.new
+  end
+
+  def create
+    @user = User.new(user_params) # Not the final implementation!
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove () and comment

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    if @user.save
+      flash[:success] = "Welcome to the Sample App!"
+      redirect_to @user
+    else
+      render :new
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb:17]: add flash for this case

File: app/views/users/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      <%= f.label :name %>
+      <%= f.text_field :name %>
+      <%= f.label :email %>
+      <%= f.email_field :email %>
+      <%= f.label :password %>
+      <%= f.password_field :password %>
+      <%= f.label :password_confirmation, "Password confirmation" %>
+      <%= f.password_field :password_confirmation %>
+      <%= f.submit "Create My Account", class: "btn btn-primary" %>
+    <% end %>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]:
```suggestion
      <%= form_for @user, url: signup_path do |f| %>
      
        <% if @user.errors.any? %>
          <div id="error_explanation">
            <div class="alert alert-danger">
              The form contains
              <%= pluralize @user.errors.count, "error" %>
            </div>
            <ul>
              <% @user.errors.full_messages.each do |msg| %>
                <li><%= msg %></li>
              <% end %>
            </ul>
          </div>
        <% end %>
      
        <%= f.label :name %>
        <%= f.text_field :name %>
        <%= f.label :email %>
        <%= f.email_field :email %>
        <%= f.label :password %>
        <%= f.password_field :password %>
        <%= f.label :password_confirmation, "Password confirmation" %>
        <%= f.password_field :password_confirmation %>
        <%= f.submit "Create My Account", class: "btn btn-primary" %>
      <% end %>
```

File: config/database.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  socket: /var/run/mysqld/mysqld.sock    
+
+development:
+  <<: *default
+  database: sample_app_development
+
+test:
+  <<: *default
+  database: sample_app_test
+
```
Comment (saobv-2223) [inline config/database.yml]: image

File: config/routes.rb [hunk 7]
```diff
@@ -1,6 +1,10 @@
 Rails.application.routes.draw do
+  get 'users/show'
```
Comment (saobv-2223) [inline config/routes.rb]: using "

File: config/settings.yml [hunk 8]
```diff
@@ -3,3 +3,13 @@ links:
...
+
+validations:
+  name:
+    max_length: 50
+  email:
+    max_length: 50
+    regex: !ruby/regexp '/\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i'
+  password:
+    min_length: 6
+
```
Comment (saobv-2223) [inline config/settings.yml]: add endline, please

**Comment by baaohuii** [comment]
ready

File: app/views/users/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        <%= f.label :name %>
+        <%= f.text_field :name %>
+        <%= f.label :email %>
+        <%= f.email_field :email %>
+        <%= f.label :password %>
+        <%= f.password_field :password %>
+        <%= f.label :password_confirmation, "Password confirmation" %>
+        <%= f.password_field :password_confirmation %>
+        <%= f.submit "Create My Account", class: "btn btn-primary" %>
+      <% end %>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]: remove 2 spaces for those lines

File: config/database.yml.example [hunk 10]
```diff
@@ -1,25 +1,16 @@
...
-# Do not set this db to the same as development or production.
 test:
   <<: *default
-  database: db/test.sqlite3
-
-production:
-  <<: *default
-  database: db/production.sqlite3
+  database: sample_app_test
+
```
Comment (saobv-2223) [inline config/database.yml.example]: please, don't change code in this file.
You only change file `database.yml`

File: config/routes.rb [hunk 11]
```diff
@@ -1,6 +1,10 @@
 Rails.application.routes.draw do
+  get "users/show"
```
Comment (saobv-2223) [inline config/routes.rb]: move it below root

**Comment by baaohuii** [comment]
ready

File: app/views/layouts/application.html.erb [hunk 12]
```diff
@@ -10,9 +10,12 @@
...
+    <div class="container">
+      <% flash.each do |message_type, message| %>
+        <div class="alert alert-<%= message_type %>"><%= message %></div>
+      <% end %>
     <%= yield %>
-   </div>
-  </body>
- </html>
+  </div>
+</body>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb]:
```suggestion
    </div>
  </body>
```

File: config/database.yml.example [hunk 13]
```diff
@@ -1,25 +1,22 @@
...
-  database: db/test.sqlite3
+  database: name_DB_test
 
 production:
   <<: *default
-  database: db/production.sqlite3
+  database: name_DB_production
+  username: username
+  password: pass
+
```
Comment (saobv-2223) [inline config/database.yml.example]: add endline

**Comment by baaohuii** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**